From 78e8228a0c9da3898e3728294e99766295870f4b Mon Sep 17 00:00:00 2001
From: Liu Linhong <linhong.liu@intel.com>
Date: Wed, 11 Mar 2015 17:53:43 +0800
Subject: [PATCH] Add otg descriptor to usb configuration desc

In USB B-OTG compliance test, host side need read otg descriptor
from DUT's configuration descriptor to enumerate DUT.
The compliance will get otg descriptor twice. This is the first
time.

Change-Id: I6ac36103b9839a9bbbe5f6970a7201ad24b5fa68
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-9620
Signed-off-by: Liu Linhong <linhong.liu@intel.com>
---
 drivers/usb/gadget/android.c    | 14 ++++++++++++++
 drivers/usb/phy/phy-intel-cht.c |  1 +
 include/uapi/linux/usb/ch9.h    |  1 +
 3 files changed, 16 insertions(+)

diff --git a/drivers/usb/gadget/android.c b/drivers/usb/gadget/android.c
index 7e9e48357487..fc090fcfcb43 100644
--- a/drivers/usb/gadget/android.c
+++ b/drivers/usb/gadget/android.c
@@ -147,6 +147,17 @@ static struct usb_device_descriptor device_desc = {
 	.bNumConfigurations   = 1,
 };
 
+static struct usb_otg_descriptor otg_descriptor = {
+	.bLength              = sizeof(otg_descriptor),
+	.bDescriptorType      = USB_DT_OTG,
+	.bcdOTG               = 0x0200, /* version 2.0 */
+};
+
+const struct usb_descriptor_header *otg_desc[] = {
+	(struct usb_descriptor_header *) &otg_descriptor,
+	NULL,
+};
+
 static struct usb_configuration android_config_driver = {
 	.label		= "android",
 	.unbind		= android_unbind_config,
@@ -1476,6 +1487,9 @@ static int android_bind_config(struct usb_configuration *c)
 	struct android_dev *dev = _android_dev;
 	int ret = 0;
 
+	if (gadget_is_otg(c->cdev->gadget))
+		c->descriptors = otg_desc;
+
 	ret = android_bind_enabled_functions(dev, c);
 	if (ret)
 		return ret;
diff --git a/drivers/usb/phy/phy-intel-cht.c b/drivers/usb/phy/phy-intel-cht.c
index 13227029e551..5919629a8e24 100644
--- a/drivers/usb/phy/phy-intel-cht.c
+++ b/drivers/usb/phy/phy-intel-cht.c
@@ -225,6 +225,7 @@ static int cht_otg_set_peripheral(struct usb_otg *otg,
 		return -EINVAL;
 
 	otg->gadget = gadget;
+	otg->gadget->is_otg = 1;
 
 	otg_dev->fsm.b_bus_req = 1;
 
diff --git a/include/uapi/linux/usb/ch9.h b/include/uapi/linux/usb/ch9.h
index aa33fd1b2d4f..3b255d9ce3e5 100644
--- a/include/uapi/linux/usb/ch9.h
+++ b/include/uapi/linux/usb/ch9.h
@@ -672,6 +672,7 @@ struct usb_otg_descriptor {
 	__u8  bDescriptorType;
 
 	__u8  bmAttributes;	/* support for HNP, SRP, etc */
+	__le16 bcdOTG;		/* release number, i.e, 2.0 is 0x0200 */
 } __attribute__ ((packed));
 
 /* from usb_otg_descriptor.bmAttributes */
-- 
2.1.0

